`Scala3RunTime$.assertFailed` w/ `-Ykind-projector:underscores`

## Compiler version

`3.1.2`, `3.2.0-RC1-bin-20220414-ba45d46-NIGHTLY`

## Minimized code

```Scala
type LiftP[F[_], T] <: Tuple =
  T match {
    case _ *: _ => F[Head[T]] *: LiftP[F, Tail[T]]
    case _ => EmptyTuple
  }
```

## Output (click arrow to expand)
<details>

```scala
$ scalac -Ykind-projector:underscores src/main/scala/bug.scala 
exception occurred while typechecking src/main/scala/bug.scala
exception occurred while compiling src/main/scala/bug.scala
java.lang.AssertionError: assertion failed while compiling src/main/scala/bug.scala
Exception in thread "main" java.lang.AssertionError: assertion failed
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
        at dotty.tools.dotc.typer.Typer$$anon$3.transform(Typer.scala:1647)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1493)
        at scala.collection.immutable.List.mapConserve(List.scala:472)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1493)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1431)
        at dotty.tools.dotc.typer.Typer$$anon$3.transform(Typer.scala:1644)
        at dotty.tools.dotc.typer.Typer.caseRest$2(Typer.scala:1715)
        at dotty.tools.dotc.typer.Typer.typedTypeCase(Typer.scala:1722)
        at dotty.tools.dotc.typer.Typer.$anonfun$39(Typer.scala:2027)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:92)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:108)
        at dotty.tools.dotc.typer.Typer.typedMatchTypeTree(Typer.scala:2027)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2838)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2872)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2937)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2941)
        at dotty.tools.dotc.typer.Typer.typedType(Typer.scala:3060)
        at dotty.tools.dotc.typer.Namer.typedAheadType$$anonfun$1(Namer.scala:1470)
        at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1463)
        at dotty.tools.dotc.typer.Namer.typedAheadType(Namer.scala:1470)
        at dotty.tools.dotc.typer.Namer$TypeDefCompleter.typeSig(Namer.scala:1013)
        at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:914)
        at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:806)
        at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:168)
        at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:188)
        at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:190)
        at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:370)
        at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2751)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2776)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2871)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2937)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2941)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2963)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3013)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2454)
        at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2797)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2801)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2871)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2937)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2941)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2963)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3013)
        at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2581)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2842)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2872)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2937)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2941)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3057)
        at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:47)
        at dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:411)
        at dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:54)
        at dotty.tools.dotc.typer.TyperPhase.runOn$$anonfun$3(TyperPhase.scala:88)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:88)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:259)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:270)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:278)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:287)
        at dotty.tools.dotc.Run.compileSources(Run.scala:220)
        at dotty.tools.dotc.Run.compile(Run.scala:204)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
        at dotty.tools.dotc.Driver.process(Driver.scala:199)
        at dotty.tools.dotc.Driver.process(Driver.scala:167)
        at dotty.tools.dotc.Driver.process(Driver.scala:179)
        at dotty.tools.dotc.Driver.main(Driver.scala:209)
        at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -1988,7 +1988,7 @@ object Parsers {
       if isSimpleLiteral then
         SingletonTypeTree(simpleLiteral())
       else if in.token == USCORE then
-        if ctx.settings.XkindProjector.value == "underscores" then
+        if ctx.settings.XkindProjector.value == "underscores" && !inMatchPattern then
           val start = in.skipToken()
           Ident(tpnme.USCOREkw).withSpan(Span(start, in.lastOffset, start))
         else

--- a/tests/pos/14952.scala
+++ b/tests/pos/14952.scala
@@ -1,0 +1,9 @@
+//> using options -Xkind-projector:underscores
+
+import Tuple.*
+
+type LiftP[F[_], T] <: Tuple =
+  T match {
+    case _ *: _ => F[Head[T]] *: LiftP[F, Tail[T]]
+    case _ => EmptyTuple
+  }

--- a/tests/pos/21400.scala
+++ b/tests/pos/21400.scala
@@ -1,0 +1,7 @@
+//> using options -Xkind-projector:underscores
+
+import scala.compiletime.ops.int.S
+
+type IndexOf[T <: Tuple, E] <: Int = T match
+  case E *: _  => 0
+  case _ *: es => 1 // S[IndexOf[es, E]]

--- a/tests/pos/21400b.scala
+++ b/tests/pos/21400b.scala
@@ -1,0 +1,10 @@
+//> using options -Xkind-projector:underscores
+
+import scala.quoted.Type
+import scala.quoted.Quotes
+
+def x[A](t: Type[A])(using Quotes): Boolean = t match
+  case '[_ *: _] =>
+    true
+  case _ =>
+    false

